### PR TITLE
Fix #5157 qcall.auth_db holds sqlalchemy session

### LINKED ISSUE #5157: 20221129 auth_db sessions on qcall

### PR DESCRIPTION
**Pull request by robnagler**
- Significant change to the auth_db interface
- Removes all use of flask.g and other global state
- centralized commit/rollback decision in uri_router. Two exceptions: nested quests and spa_session

### CODE REVIEW

**Comment by robnagler** [comment]
@e-carlin I switched reviewer to you from @moellep. This is somewhat high priority as it changes 28 files, some pretty intensely.

**Comment by e-carlin** [comment]
Ok, I'll review today.

**Comment by moellep** [comment]
I manually tested this branch with email, guest, github and bluesky auth and everything worked OK. @e-carlin, please make sure oauth and ldap login still works.

File: sirepo/auth/__init__.py [hunk 1]
```diff
@@ -82,11 +82,11 @@
 _cfg = None
 
 
-def init_quest(qcall):
+def init_quest(qcall, top_level_call_api=False):
```
Comment (e-carlin) [inline sirepo/auth/__init__.py:85]: I think it is worth documenting this method. Not exactly clear to me what top_level_call_api is and why this init_quest needs to be different from all of the others.

File: sirepo/auth/__init__.py [hunk 2]
```diff
@@ -219,22 +219,34 @@ def cookie_cleaner(self, values):
         values[_COOKIE_STATE] = _STATE_LOGGED_OUT
         return values
 
-    def create_user(self, module, want_login=False):
+    def create_user(self, module_or_method, want_login=False):
```
Comment (e-carlin) [inline sirepo/auth/__init__.py:222]: Looks to me that module_or_method is only ever a method

File: sirepo/auth/__init__.py [hunk 3]
```diff
@@ -219,22 +219,34 @@ def cookie_cleaner(self, values):
...
+        m = _METHOD_MODULES.get(module_or_method, module_or_method)
         u = simulation_db.user_create()
         if want_login:
-            self._login_user(module, u)
-            self._create_roles_for_new_user(module.AUTH_METHOD)
+            self._login_user(m, u)
+            self._create_roles_for_new_user(m.AUTH_METHOD)
```
Comment (e-carlin) [inline sirepo/auth/__init__.py:227]: m.AUTH_METHOD won't work if m is a method and not a module.

File: sirepo/auth/__init__.py [hunk 4]
```diff
@@ -531,7 +542,9 @@ def reset_state(self):
 
     def unchecked_get_user(self, uid):
         with sirepo.util.THREAD_LOCK:
-            u = sirepo.auth_db.UserRegistration.search_by(uid=uid)
+            u = self.qcall.auth_db.model("UserRegistration").unchecked_search_by(
```
Comment (e-carlin) [inline sirepo/auth/__init__.py:545]: Really nice to be moving away from globals

**Review by e-carlin — APPROVED** [review]
Couple nits but lgtm

**Comment by e-carlin** [comment]
@moellep thanks for testing. Confirmed oauth works. LDAP will be added soon.

File: sirepo/auth/__init__.py (hunk 1 above)
Comment (robnagler) [reply sirepo/auth/__init__.py:85]: I documented to show that it was unpublished 695c9ef3f83ac271fecd1f4f45f84129ddb26239

File: sirepo/auth/__init__.py (hunk 3 above)
Comment (robnagler) [reply sirepo/auth/__init__.py]: Was a mess, made private 695c9ef3f83ac271fecd1f4f45f84129ddb26239

File: sirepo/auth/__init__.py (hunk 2 above)
Comment (robnagler) [reply sirepo/auth/__init__.py]: login is a bit goofy. It said "method" but was a module so I renamed a var and took create_user private. login needs to be broken up. 695c9ef3f83ac271fecd1f4f45f84129ddb26239

File: sirepo/auth/__init__.py (hunk 4 above)
Comment (robnagler) [reply sirepo/auth/__init__.py:539]: Yup

**Pull request merged**